### PR TITLE
nix-collect-garbage: Make the --dry-run option work, as long as no generations are being deleted.

### PR DESCRIPTION
**Pull request by DavidEGrayson**
This is the first step of fixing issue #1751: it makes `nix-collect-garbage --dry-run` work.  It also prints an error if `--dry-run` is used in conjunction with the options for deleting generations, because we don't have the machinery needed to print which outputs reachable from those old generations are going to be deleted.

The code in this commit is similar to [lines 599-604 of nix-store.cc](https://github.com/NixOS/nix/blob/5f1891b/src/nix-store/nix-store.cc#L599-L604).

CC: @shlevy

### CODE REVIEW

File: src/nix-collect-garbage/nix-collect-garbage.cc [hunk 1]
```diff
@@ -76,19 +76,31 @@ static int _main(int argc, char * * argv)
             return true;
         });
 
+        if (removeOld && dryRun)
+        {
+            // removeOldGenerations does not record any information
+            // about which generations would be deleted if dryRun is true, so
+            // we cannot get an accurate list of store paths to be deleted.
+            std::cerr << "error: Sorry, the --dry-run option is not yet "
+                "compatible with deleting generations." << std::endl;
```
Comment (edolstra) [inline src/nix-collect-garbage/nix-collect-garbage.cc]: This should be `throw UsageError("bla bla");`.

Comment (DavidEGrayson) [reply src/nix-collect-garbage/nix-collect-garbage.cc]: Thanks for taking a look!  I amended my commit so it uses `UsageError` now.

**Comment by bew** [comment]
Hello!

I've juste been hit by #1751 today (as a new nix user), I see there has not been any work on this since Feb 2019.

@DavidEGrayson I hope you're doing well in the current world situation, are you going to continue working on this subject?

Thanks!

**Comment by DavidEGrayson** [comment]
This pull request has been ignored for over a year. I'm not really interested in doing more work if my previous work is ignored.   I'm glad you at least see the value in it, @bew.

**Pull request closed without merging**